test/builtin_dataset_mocks.py

```diff
@@ -62,27 +62,47 @@ def _parse_mock_info(self, mock_info):
 
         return mock_info
 
-    def prepare(self, config):
+    def load(self, config):
         # `datasets.home()` is patched to a temporary directory through the autouse fixture `test_home` in
         # test/test_prototype_builtin_datasets.py
         root = pathlib.Path(datasets.home()) / self.name
-        root.mkdir(exist_ok=True)
+        mock_data_folder = root / "__mock__"
+        mock_data_folder.mkdir(parents=True)
 
-        mock_info = self._parse_mock_info(self.mock_data_fn(root, config))
+        mock_info = self._parse_mock_info(self.mock_data_fn(mock_data_folder, config))
 
-        with unittest.mock.patch.object(datasets.utils.Dataset, "__init__"):
-            required_file_names = {
-                resource.file_name for resource in datasets.load(self.name, root=root, **config)._resources()
-            }
-        available_file_names = {path.name for path in root.glob("*")}
-        missing_file_names = required_file_names - available_file_names
-        if missing_file_names:
-            raise pytest.UsageError(
-                f"Dataset '{self.name}' requires the files {sequence_to_str(sorted(missing_file_names))} "
-                f"for {config}, but they were not created by the mock data function."
-            )
+        def mock_data_download(resource, root, **kwargs):
+            src = mock_data_folder / resource.file_name
+            if not src.exists():
+                raise pytest.UsageError(
+                    f"Dataset '{self.name}' requires the file {resource.file_name} for {config}"
+                    f"but it was not created by the mock data function."
+                )
 
-        return mock_info
+            dst = root / resource.file_name
+            shutil.move(str(src), str(root))
+
+            return dst
+
+        with unittest.mock.patch(
+            "torchvision.prototype.datasets.utils._resource.OnlineResource.download", new=mock_data_download
+        ):
+            dataset = datasets.load(self.name, **config)
+
+        extra_files = list(mock_data_folder.glob("**/*"))
+        if not extra_files:
+            mock_data_folder.rmdir()
+        else:
+            pass
+            # raise pytest.UsageError(
+            #     (
+            #         f"Dataset '{self.name}' created the following files for {config} in the mock data function, "
+            #         f"but they were not loaded:\n\n"
+            #     )
+            #     + "\n".join(str(file.relative_to(mock_data_folder)) for file in extra_files)
+            # )
```

Comment:
This was not enforced before and a few datasets are failing. I already fixed this in the open PR #5706. If this PoC is approved, I can migrate them here.